### ISSUE TITLE
Metric for Size of code base

### ISSUE THREAD

**Issue by elong0527** [pharmaR/riskmetric#66]
"The larger the code base, the greater the risk of error" from https://github.com/pharmaR/pharmaR.github.io/blob/master/content/risk.md

I suggest to remove this metric. The size can be defined in different ways. If it is the size of an CRAN version, it should be <= 5MB per CRAN package size and limitation.

**Comment by elong0527** [comment]
How to define the metric? 

option 1: source code size 
    -- if source code is not available in local. How could we access the same number from remote. 

option 2: number of R functions (internal vs. external) 

option 3: number of lines of R code. 

option 4: Overal lines of R code divided by number of R functions

**Comment by mpadge** [comment]
This is definitely a core component of what we'll be working on, so should be able to constructively collaborate here. I would hope that, rather than
> remove this metric

it should be possible to work towards one or more metrics that are both more specific and more useful. Might be useful to know that I've run a [`cloc`](https://github.com/hrbrmstr/cloc) on the entire CRAN archive of nearly 16,000 packages in just a few minutes, so computational complexity of metrics ought not really be any issue here. One of our ideas is to quantify a bunch of these kinds of metrics *for all extant packages* and to use them to flag packages that might be outliers in the total multi-dimensional space (along with info on the dimensions most contributing to outlier status). Rather than a single metric, "risk" (in your context) then naturally could then naturally emerge from a bunch of metrics which could be automatically dumped in an assessment and only flagged in the final `score` phase, effectively and neatly avoiding any direct need to decide which single metric(s) to "use".

Our approach -- and i'm sure yours too -- is going to be to defer any decisions regarding implementation of final `score` systems as long as practicable, and work on `assess` steps with often little more than a qualitative or instinctive feel for how these are going to contribute to `score` outputs. I'm optimistic that this approach in this particular context is going to work quite well.

**Comment by elong0527** [comment]
I agree to focus on the metrics to be collected for now. Shall we collect the information in the R folder as below?

*  number of R files
*  number of lines in each R file
*  number of internal R functions in each R file
*  number of external R functions in each R file

**Comment by dgkf** [comment]
@elong0527 - I think we should take @mpadge's approach for this one. We could use `cloc` to get a number of coding lines. Let's hold off on the scoring - I think it's really hard to define a way of scoring lines of code holistically. Perhaps the research @mpadge is doing can provide some guidelines based on characteristics of those outliers.

I think that this one probably relies on a few other characteristics of a package to really build a complete picture. I could see something like a ratio between lines of code and exported functions being a better measure of quality than either individually.

**Comment by elong0527** [comment]
@paulie-of-punskas this might be a good item to contribute if you are interested in.

**Comment by shengwei66** [comment]
I am exploring the `cloc` package, and found that the `cloc_cran` function can easily create a tibble with the number of R files (file_count), lines of code (loc), blank_lines, and comment_lines. So this function can get the metric for the pkg_cran_remote class. However, for the pkg_source and pkg_install class, it's not clear to me about how to use the `cloc` function for getting these numbers.

**Comment by elong0527** [comment]
- one possible package to check is the `covr` package that shall have a way to calculate number of lines.

**Comment by mpadge** [comment]
I'd suggest that the `cloc` package isn't really what you want here, as the following code demonstrates:

``` r
library (pkgstats)
packageVersion ("pkgstats")
#> [1] '0.0.3.88'
u <- "https://github.com/pharmaR/riskmetric"
path <- file.path (tempdir (), "riskmetric")
gert::git_clone (u, path = path)
s <- pkgstats (path)
s$loc
#> # A tibble: 4 × 12
#> # Groups:   language, dir [4]
#>   language dir       nfiles nlines ncode  ndoc nempty nspaces nchars nexpr ntabs
#>   <chr>    <chr>      <int>  <int> <int> <int>  <int>   <int>  <int> <dbl> <int>
#> 1 HTML     tests          4    549     0   520     29    3522  47007    NA     0
#> 2 R        R             67   3901  1703  1561    637   11742  99322     2     0
#> 3 R        tests         19    533   363    85     85    1173  15646     1     0
#> 4 Rmd      vignettes      2    606   433    30    143    2676  18682     1     0
#> # … with 1 more variable: indentation <int>
cloc::cloc (path)
#> # A tibble: 8 × 10
#>   source     language file_count file_count_pct   loc  loc_pct blank_lines
#>   <chr>      <chr>         <int>          <dbl> <int>    <dbl>       <int>
#> 1 riskmetric R                84        0.416    2060 0.328            722
#> 2 riskmetric HTML              4        0.0198    517 0.0822            29
#> 3 riskmetric JSON              1        0.00495   224 0.0356             6
#> 4 riskmetric YAML              4        0.0198    127 0.0202            26
#> 5 riskmetric Markdown          5        0.0248    109 0.0173            43
#> 6 riskmetric Rmd               2        0.00990   106 0.0169           143
#> 7 riskmetric SVG               1        0.00495     1 0.000159           0
#> 8 riskmetric SUM             101        0.5      3144 0.5              969
#> # … with 3 more variables: blank_line_pct <dbl>, comment_lines <int>,
#> #   comment_line_pct <dbl>
```

<sup>Created on 2022-02-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

`cloc` just counts code regardless of where it is, and so gives simple totals which could be from anywhere. `R` code can be in any part of `inst`, or `tests`, or `vignettes`, or any number of other places. `pkgstats` distinguishes both the language used and the sub-directory, so you can, for example, just count `R` code in the `R` directory. The internal lines-of-code routines in `pkgstats` were developed to ensure they did exactly what [the original `cloc` script](https://github.com/AlDanial/cloc) does and a whole lot more, including necessary R-specific stuff like processing `.Rmd` files to distinguish chunks, and identifying `roxygen2` documentation lines (which `cloc` can't do). That enables you to use `pkgstats` to derive metrics like proportion of documentation-to-code lines for each file, or even each function. The differences in LoC are also revealing: `cloc` gives 2060 LoC from 84 files; `pkgstats` gives 1703 from 67 files, which is more accurate.

The full data set of the entire CRAN history has now been analysed with `pkgstats`, and can be used to get LoC counts for all versions of all packages with even having to use the package itself. Data are included as part of [the `pkgstats` releases](https://github.com/ropensci-review-tools/pkgstats/releases), and will be updated monthly.

You can use that for any CRAN packages, and you'd only need to run `pkgstats` for non-CRAN packages, in which case you'd just need to develop a method for `pkg_source`. (And of course `pkg_install` objects have no associated code, so can't have a Lines-of-Code metric, but can just grab the associated metrics either from the database for CRAN packages, or from the associated remote/source path otherwise.)

**Comment by elimillera** [comment]
@shengwei66 @emilliman5  I believe this was closed in #249. Are we good to close this?
